### ISSUE TITLE
Shallow fusion of LSTM LM

### ISSUE THREAD

**Issue by akshatdewan** [rwth-i6/returnn-experiments#4]
Hi, 

I trained an LSTM LM using [this](https://github.com/rwth-i6/returnn-experiments/blob/master/2018-asr-attention/librispeech/lm/i512_m2048_m2048.sgd_b64_lr0_cl2.newbobabs.d0.2.config) config file and also an attention model using the [this](https://github.com/rwth-i6/returnn-experiments/blob/master/2018-asr-attention/librispeech/full-setup-attention/returnn.config) config. The attention model alone gives impressive (18.3 WER) results on a test set composed of accented speech from multiple speakers in a conference setting. I am using it for building an ASR system for a UN specialized agency with no intentions of commercializing it. 

But, now I am trying to figure out how I can integrate the LM using shallow fusion (as mentioned in your paper - Improved training of end-to-end..) There does not seem to be a layer_class for it like kenlm used [here](https://github.com/rwth-i6/returnn-experiments/blob/master/2018-asr-attention/librispeech/attention/exp3.ctc.lmbpe3gr.config#L117). 

Could you please help me understand if I need some special class to define in the [RecLayer](https://github.com/rwth-i6/returnn/blob/17018149bc19540e7b8857054572eb1128a733a0/TFNetworkRecLayer.py) file or can I do it differently? 

Thanks for your great work. 

Akshat

**Comment by albertz** [comment]
@kazuki-irie Can you upload your shallow fusion config also ([here](https://github.com/rwth-i6/returnn-experiments/tree/master/2018-asr-attention/librispeech/attention))?

@akshatdewan Yes, sorry, we forgot to upload that config. You will find a similar config for shallow fusion for Switchboard [here](https://github.com/rwth-i6/returnn-experiments/blob/master/2018-asr-attention/switchboard/attention/multihead-mlp-h1.red8.enc6l.encdrop03.decbs.ls01.pretrain2.nbd07.lm.config). It should not be hard to adapt this in the same way for LibriSpeech.

**Comment by akshatdewan** [comment]
Thanks @albertz I will try modifying the SWB config file.

**Comment by kazuki-irie** [comment]
I have added the config.

**Comment by albertz** [comment]
Thanks! I'm closing this now. Feel free to reopen if there are further issues.

**Comment by akshatdewan** [comment]
Thank you! I just have a question regarding the input to the lm subnetwork - 
I am having troubles understanding what it means when it says `"from":["prev:output"]` [here](https://github.com/rwth-i6/returnn-experiments/blob/master/2018-asr-attention/librispeech/attention/exp3.ctc.lm.config#L132). Could you please help me understand where the input to the layer `lm_out` is coming from? 
Thanks

**Comment by albertz** [comment]
It means that the input to the LM is the output of the previous frame.

**Comment by akshatdewan** [comment]
Thanks! And I am sorry, I missed it [here](https://github.com/rwth-i6/returnn/blob/17018149bc19540e7b8857054572eb1128a733a0/TFNetworkRecLayer.py#L39) 
`In the subnetwork, you can access outputs from layers from the previous time step when they
  are referred to with the "prev:" prefix.`

**Comment by yanghongjiazheng** [comment]
@akshatdewan 
Hi,

I wang to train an LSTM LM using the config. But I have no idea about the exactly form of the /work/asr3/irie/data/librispeech/lm_bpe/librispeech-lm-norm.bpe.txt.gz . So I can not shape my data. And also I coming to the trouble of  **def cf(filename)** , err_msg = <local> "No such file or directory: 'cf'".  
Can you please help me out.
Thanks for your help.

**Comment by albertz** [comment]
@yanghongjiazheng See my answer in #21, and please let's keep any further discussion/question in there.
